test/e2e/gsm-secret-sync/gsm-e2e_test.go

```diff
@@ -0,0 +1,495 @@
+	expectedState, err := getExpectedState(configPath, tr.config)
+	if err != nil {
+		t.Fatalf("failed to get expected state from config %s: %v", configPath, err)
+	}
+	err = tr.runReconcilerTool(configPath)
+	if err != nil {
+		t.Fatalf("failed to run reconciler tool: %v", err)
+	}
+	actualState := tr.getActualGCPStateWithRetry(expectedState)
+
+	if !tr.compareStates(actualState, expectedState) {
+		t.Fatal("update test failed")
+	}
+}
+
+func TestDeletion(t *testing.T) {
+	configPath := "testdata/config-delete.yaml"
+	expectedState, err := getExpectedState(configPath, tr.config)
+	if err != nil {
+		t.Fatalf("failed to get expected state from config %s: %v", configPath, err)
+	}
+	err = tr.runReconcilerTool(configPath)
+	if err != nil {
+		t.Fatalf("failed to run reconciler tool: %v", err)
+	}
+	actualState := tr.getActualGCPStateWithRetry(expectedState)
+
+	if !tr.compareStates(actualState, expectedState) {
+		t.Fatal("deletion test failed")
+	}
+}
+
+// runReconcilerTool runs the reconciler tool's binary with the given config path
+func (tr *testRunner) runReconcilerTool(configPath string) error {
+	credFile := os.Getenv("GOOGLE_APPLICATION_CREDENTIALS")
+	cmd := exec.Command(tr.binaryPath,
+		"--config", configPath,
+		"--log-level", "info",
+		"--gcp-service-account-key-file", credFile)
+
+	cmd.Env = append(os.Environ(),
+		fmt.Sprintf("GCP_PROJECT_ID=%s", tr.config.ProjectIdString),
+		fmt.Sprintf("GCP_PROJECT_NUMBER=%s", tr.config.ProjectIdNumber),
+	)
+
+	output, err := cmd.CombinedOutput()
+	if err != nil {
+		return fmt.Errorf("google secret manager sync failed: %w\nOutput:\n%s", err, string(output))
+	}
+	logrus.Info("Running reconciler tool")
+	if *logLevel == "debug" {
+		fmt.Print(string(output))
+	}
+	return nil
+}
+
+// getActualGCPState returns the current state of resources in the GCP project
+func (tr *testRunner) getActualGCPState() GCPState {
+	secrets, err := gsm.GetAllSecrets(tr.ctx, tr.secretsClient, tr.config)
+	if err != nil {
+		logrus.Errorf("Error while fetching secrets: %v", err)
+		secrets = make(map[string]gsm.GCPSecret)
+	}
+
+	serviceAccounts, err := gsm.GetUpdaterServiceAccounts(tr.ctx, tr.iamAdminClient, tr.config)
+	if err != nil {
+		logrus.Errorf("Failed to fetch service accounts: %v", err)
+		serviceAccounts = []gsm.ServiceAccountInfo{}
+	}
+
+	var iamBindings []*iampb.Binding
+	policy, err := gsm.GetProjectIAMPolicy(tr.ctx, tr.resourceManagerClient, tr.config.ProjectIdNumber)
+	if err != nil {
+		logrus.Errorf("Failed to fetch IAM policy: %v", err)
+	} else {
+		for _, binding := range policy.Bindings {
+			if gsm.IsManagedBinding(binding) {
+				iamBindings = append(iamBindings, binding)
+			}
+		}
+	}
+
+	return GCPState{
+		Secrets:         secrets,
+		IAMBindings:     iamBindings,
+		ServiceAccounts: serviceAccounts,
+	}
+}
+
+// getActualGCPStateWithRetry gets the actual state with retry for eventual consistency
+func (tr *testRunner) getActualGCPStateWithRetry(expectedState GCPState) GCPState {
```

Comment:
Not sure whether should I raise this concern or not as it might not be a big deal. I would expect `getActualGCPStateWithRetry` to read the state from GCP and nothing else, instead of reading the state and asserting something (the number of SA and secrets in this case).

Reply:
Yeah, I see your point. My reasoning for this is that because GCP sometimes takes a while to propagate changes, so if I see that I'm missing some resources from the expected state, I do a few retries to make sure the difference is not because of GCP's slowness, otherwise I'd get a lot of false negatives.